File: src/main/java/GuessNumberHigherOrLower/Fanrencli/374.guess-number-higher-or-lower.java
/*
 * @lc app=leetcode id=374 lang=java
 *
 * [374] Guess Number Higher or Lower
 */
/* The guess API is defined in the parent class GuessGame.
   @param num, your guess
   @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
      int guess(int num); */

public class Solution extends GuessGame {
    int temp;
    int num;
    public int guessNumber(int n) {
        return checkResult(1,n);
    }
    public int checkResult(int start, int end){
        num=start+(end-start)/2;
        temp=guess(num);
        if(temp==1 & num==start){
            return end;
        }
        if(temp==1){
            return checkResult(num+1,end);
        }else if(temp==-1){
            return checkResult(start,num-1);
        }else{
            return num;
        }
    }
}

File: src/main/java/GuessNumberHigherOrLower/GGOBOND/374_Guess Number Higher or Lower.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
# def guess(num):

class Solution(object):
    def guessNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        """
        AC:
        与278. 第一个错误的版本这道题作对比！！！（同样是二分法）
        1.创建头尾两指针l，r
        2.mid，判断mid对应情况，从而决定l和r指针向哪个方向移动
        NOTE: 3. while l <= r: (此处应该考虑等号)-->考虑当n==1时的情况，即l==r==1
        """

        l, r = 1, n
        while l <= r:  # 考虑等号
            mid = (l + r) // 2
            cur_res = guess(mid)
            print(mid, cur_res)
            if cur_res == -1:
                if guess(mid - 1) == 0:
                    return mid - 1
                r = mid - 1
            elif cur_res == 1:
                if guess(mid + 1) == 0:
                    return mid + 1
                l = mid + 1
            elif cur_res == 0:
                return mid
